clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -347,16 +352,35 @@ private void sendProduceRequests(Map<Integer, List<RecordBatch>> collated, long
      * Create a produce request from the given record batches
      */
     private void sendProduceRequest(long now, int destination, short acks, int timeout, List<RecordBatch> batches) {
+        if (batches.isEmpty())
+            return;
+
         Map<TopicPartition, MemoryRecords> produceRecordsByPartition = new HashMap<>(batches.size());
         final Map<TopicPartition, RecordBatch> recordsByPartition = new HashMap<>(batches.size());
+
+        // find the minimum magic version used when creating the record sets
+        byte minUsedMagic = apiVersions.maxUsableProduceMagic();
+        for (RecordBatch batch : batches) {
+            if (batch.magic() < minUsedMagic)
+                minUsedMagic = batch.magic();
+        }
+
         for (RecordBatch batch : batches) {
             TopicPartition tp = batch.topicPartition;
-            produceRecordsByPartition.put(tp, batch.records());
+            MemoryRecords records = batch.records();
+
+            // down convert if necessary to the minimum magic used. This is intended to handle edge cases around
+            // cluster upgrades where brokers may not all support the same message format version. For example,
+            // if a partition migrates from a broker which is supporting the new magic version to one which doesn't,
+            // then we will need to convert.
```

Comment:
Hmm, this adds some complexity and I am not sure about the benefit. First, when sending the first few batches of the data in the producer or when the leaders are not on all brokers, we don't have a connection to all brokers. So the minUsedMagic may not be very accurate.

Could we just rely on the down conversion logic on the broker side and make it clearer in the documentation that EOS feature should be turned on only when the whole cluster has been upgraded?